frame/staking/src/pallet/impls.rs

```diff
@@ -629,54 +639,79 @@ impl<T: Config> Pallet<T> {
 
 	/// Get all of the voters that are eligible for the npos election.
 	///
-	/// This will use all on-chain nominators, and all the validators will inject a self vote.
+	/// `voter_count` imposes a cap on the number of voters returned; care should be taken to ensure
+	/// that it is accurate.
+	///
+	/// This will use on-chain nominators, and all the validators will inject a self vote.
 	///
 	/// This function is self-weighing as [`DispatchClass::Mandatory`].
 	///
 	/// ### Slashing
 	///
 	/// All nominations that have been submitted before the last non-zero slash of the validator are
 	/// auto-chilled.
-	pub fn get_npos_voters() -> Vec<(T::AccountId, VoteWeight, Vec<T::AccountId>)> {
-		let weight_of = Self::slashable_balance_of_fn();
-		let mut all_voters = Vec::new();
+	pub fn get_npos_voters(
+		maybe_max_len: Option<usize>,
+	) -> Vec<(T::AccountId, VoteWeight, Vec<T::AccountId>)> {
+		let nominator_count = CounterForNominators::<T>::get() as usize;
+		let validator_count = CounterForValidators::<T>::get() as usize;
+		let all_voter_count = validator_count.saturating_add(nominator_count);
+		drop(validator_count);
+		drop(nominator_count);
 
-		let mut validator_count = 0u32;
-		for (validator, _) in <Validators<T>>::iter() {
+		let max_allowed_len = maybe_max_len.unwrap_or(all_voter_count).min(all_voter_count);
+		drop(all_voter_count);
+
+		let mut all_voters = Vec::<_>::with_capacity(max_allowed_len);
+
+		// first, grab all validators, capped by the maximum allowed length.
+		let mut validators_taken = 0u32;
+		for (validator, _) in <Validators<T>>::iter().take(max_allowed_len) {
 			// Append self vote.
-			let self_vote = (validator.clone(), weight_of(&validator), vec![validator.clone()]);
+			let self_vote =
+				(validator.clone(), Self::weight_of(&validator), vec![validator.clone()]);
 			all_voters.push(self_vote);
-			validator_count.saturating_inc();
+			validators_taken.saturating_inc();
 		}
 
-		// Collect all slashing spans into a BTreeMap for further queries.
+		// .. and grab whatever we have left from nominators.
+		let nominators_quota = max_allowed_len.saturating_sub(validators_taken as usize);
 		let slashing_spans = <SlashingSpans<T>>::iter().collect::<BTreeMap<_, _>>();
-
-		let mut nominator_count = 0u32;
-		for (nominator, nominations) in Nominators::<T>::iter() {
-			let Nominations { submitted_in, mut targets, suppressed: _ } = nominations;
-
-			// Filter out nomination targets which were nominated before the most recent
-			// slashing span.
-			targets.retain(|stash| {
-				slashing_spans
-					.get(stash)
-					.map_or(true, |spans| submitted_in >= spans.last_nonzero_slash())
-			});
-
-			if !targets.is_empty() {
-				let vote_weight = weight_of(&nominator);
-				all_voters.push((nominator, vote_weight, targets));
-				nominator_count.saturating_inc();
+		let mut nominators_taken = 0u32;
+		for nominator in T::SortedListProvider::iter().take(nominators_quota) {
```

Comment:
the take here is a bit inaccurate as if a nominator has no vote due to slashing, then it won't be included in the snapshot.

Reply:
that's a good point. I can see it being problematic if there is a huge slahsing event that would largely affect the top n nominators (since they where likely the same voters in the last snapshot).

Assuming the `max_allowed_len` is configured with memory safety in mind, I am not sure how we could efficiently try to get more nominators if most end up being slashed. In an ideal world we could start taking from `SortedListProvider` at the index we left off at, but we don't have that ability at all in the current design. 

If this is important we could look into some design that returns the last bag that the `take` touches: if we need more nominators we can then start itrating from that bag. Its a bit of design change at this point though, especially since we have already had a succesful audit.

With all that said, from my understanding this current design basically holds up the status

Curious if @kianenigma has thoughts, and if we should maybe make an issue to modify this so it can adjust for slashed nominators and always try to ensure we reach `maybe_max_len`.

Reply:
the comment is valid and the answer is simple: this needs to change to `for x in  T::SortedListProvider::iter()` that breaks out after `nominators_quota` have been taken. In fact, we already have `nominators_taken` traking this.

Reply:
at first glance, the fix is simple; we remove the `take` and instead `break` out of the loop when `nominators_taken == nominators_quota`. This does not pose much of a risk to the memory as far as I can say, since the nominators that we have skipped have already been dropped. Maybe there should be some pessimistic bound here `nominators_quota * 2` just to prevent max heap exhaustion and a potential bug in bags-list that causes infinite loop.

Reply:
@kianenigma   14b91205b0ea014b77340f712fd90b1817402e58 I implement that plan here but still need to add tests (will do tmrw)

Reply:
looks good to me!

Reply:
@kianenigma tests e6dd68bbeab047f90b270059da856f3298e6ade4 & 2e9e31702bf751d2c009789552961d247b297f63